stablehlo/reference/Ops.cpp

```diff
@@ -33,13 +34,74 @@ namespace mlir {
 namespace stablehlo {
 namespace {
 
-Index evalIndices(ArrayRef<Tensor> runtimeIndices) {
-  Index index(runtimeIndices.size());
-  for (size_t i = 0; i < runtimeIndices.size(); ++i)
-    index[i] = runtimeIndices[i].get({}).getIntegerValue().getSExtValue();
+Index evalIndices(ArrayRef<Tensor> indices) {
+  Index index(indices.size());
+  for (size_t i = 0; i < indices.size(); ++i)
+    index[i] = indices[i].get({}).getIntegerValue().getSExtValue();
   return index;
 }
 
+Tensor evalPadOp(const Tensor &operand, const Tensor &paddingValue,
+                 const Sizes &edgePaddingLow, const Sizes &edgePaddingHigh,
+                 const Sizes &interiorPadding) {
+  SmallVector<Type> inferredTypes;
+  Builder builder(operand.getType().getContext());
+  auto inferStatus =
+      hlo::inferPadOp({}, operand.getType(), paddingValue.getType(),
+                      builder.getI64TensorAttr(edgePaddingLow),
+                      builder.getI64TensorAttr(edgePaddingHigh),
+                      builder.getI64TensorAttr(interiorPadding), inferredTypes);
+  if (failed(inferStatus))
+    report_fatal_error(invalidArgument("Could not infer PadOp's return type"));
+  return evalPadOp(operand, paddingValue, edgePaddingLow, interiorPadding,
+                   inferredTypes[0].cast<ShapedType>());
+}
+
+SmallVector<Tensor> evalReduceOp(ArrayRef<Tensor> inputs,
+                                 ArrayRef<Tensor> initValues,
+                                 const Axes &dimensions, Region &body,
+                                 Scope &scope) {
+  SmallVector<Type> inputTypes;
+  for (auto input : inputs) inputTypes.push_back(input.getType());
+
+  SmallVector<Type> initValueTypes;
+  for (auto initValue : initValues)
+    initValueTypes.push_back(initValue.getType());
+
+  SmallVector<ShapedTypeComponents> inferredReduceTypes;
+  Builder builder(inputs[0].getType().getContext());
+  auto reduceStatus = hlo::inferReduceOp(
+      /*location=*/{}, inputTypes, initValueTypes,
+      builder.getI64TensorAttr(dimensions), inferredReduceTypes);
+  if (failed(reduceStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer ReduceOp's return type"));
+
+  SmallVector<ShapedType> resultTypes;
+  for (auto inferredType : inferredReduceTypes) {
+    ShapedType shapedType;
+    if (failed(hlo::createShapedType(inferredType, shapedType)))
+      llvm::report_fatal_error("Could not retrieve type.");
```

Comment:
This error message seems a bit too cryptic. Perhaps something along the lines of "Could not infer ReduceOp's return type"?